joomla/joomla-cms
# [4.0] Show changelog - GSoC Expand Extensions Manager #17536 redo

## Redo
This is a continuation of issue #17536 as I can no longer push to that repo, starting a new PR.

### Introduction

At this time, users have to look for external sources of information about changes in an extension, when they are going to update their extensions. In most cases, users install an extension without checking the changelog due to the difficulty to find the associated changelog. With this new feature, users can review the changelog of the extensions when they are about to install an update (granted that the changelog is provided by the extension's developer), and after the extension is installed.

### Summary of Changes

Added a new parameter on the xml updateserver files 'changelogurl' and adding it to #__updates table.

### Testing Instructions

1. Apply this  pull request
2. Run the command `npm ci` to compile the JavaScript
3. Login to the administrator side
4. Go to Extensions -> Database -> Fix to make sure the database changes are applied
![screenshot from 2017-07-24 21-36-09](https://user-images.githubusercontent.com/20955884/28543880-43171938-70b9-11e7-9248-e3eff8b422d4.png)
5. Install the test extension from https://github.com/roland-d/component_joomla/releases/tag/v0.0.2
6. Go to **Manage**
7. Filter the list on `List of Students`
8. Click on the version number `0.0.2`
9. A modal shows up with the changelog as shown below under `Manage View`
10. Go to **Update**
11. Click on Find Updates
12. The update is shown and click on the **Changelog** button
13. A modal shows up with the changelog as shown below under `Update View`

### Changelog format
The changelogurl is supported for all extensions, you just need to add <changelogurl> tags in your installation xml with a valid url.
`<changelogurl>https://rolandd.com/24026/changelog.xml</changelogurl>`

The format of the XML file is as follows:
```
<changelogs>
    <changelog>
        <element>com_lists</element>
        <type>component</type>
        <version>4.0.0</version>
        <security>
            <item>Item A</item>
            <item>Item b</item>
        </security>
        <fix>
            <item>Item A</item>
            <item>Item b</item>
        </fix>
        <language>
            <item>Item A</item>
            <item>Item b</item>
        </language>
        <addition>
            <item>Item A</item>
            <item>Item b</item>
        </addition>
        <change>
            <item>Item A</item>
            <item>Item b</item>
        </change>
        <remove>
            <item>Item A</item>
            <item>Item b</item>
        </remove>
        <note>
            <item>Item A</item>
            <item>Item b</item>
        </note>
</changelog>
<changelog>
	<element>com_lists</element>
	<type>component</type>
	<version>0.0.2</version>
	<security>
		<item>Big issue</item>
	</security>
</changelog>
</changelogs>
```
This file contains 2 changelogs:
- Version 0.0.2 (for testing the manage view)
- Version 4.0.0 (for testing the update view)

 **Update View** 
Install the List component https://github.com/roland-d/component_joomla/releases/tag/v0.0.2
By default you will have always an update from 0.0.2 to 4.0.0 just for testing

If the changelog is not available it will show N/A like in the Folder column

<img width="1080" alt="image" src="https://user-images.githubusercontent.com/359377/53759695-3af11880-3ec1-11e9-911b-ea0988e66e0b.png">

It will open a modal window that loads the changelog via AJAX from the given changelog URL.
The title of the modal is version ' - ' component_name

<img width="809" alt="image" src="https://user-images.githubusercontent.com/359377/53991446-e8b33000-412a-11e9-8b5b-b9a1bfc01159.png">

**Manage View**
On the manage view the changelog is triggered by clicking on the version number.

<img width="1121" alt="image" src="https://user-images.githubusercontent.com/359377/53761061-e2237f00-3ec4-11e9-9f5b-13cd0342cdad.png">
<img width="813" alt="image" src="https://user-images.githubusercontent.com/359377/53991505-07192b80-412b-11e9-81d5-3f9849a3039a.png">


### Expected result

In Updates View you will have a new column that will show a changelog button to open a modal if there is a changelogurl in the #__updates table.
In Manage View you will have an <a> tag in the version with a info icon on the right side if your extension has a changelogurl in the #__extensions table.

### Documentation Changes Required

Add changelogurl parameter on updatesite information and in extension.
Add changelog XML format to updatesite information
